### ISSUE TITLE
[Feature Request] EnPass Support

### ISSUE THREAD

**Issue by rushsteve1** [nextcloud/passman#159]
[EnPass](https://www.enpass.io/) is a free password manager that already supports syncing via Nextcloud/ownCloud/WebDAV, so having support for synced Databases would be an amazing feature to have.

**Comment by brantje** [comment]
Can you provide an example export file? Writing an importer won't be a problem.

**Comment by rushsteve1** [comment]
EnPass exports its data as simple TXT files. Example:
```
Title : Google
Username : username
Email : username@gmail.com
Password : password123
Title : Social
Number : 123456789
Name : John Smith
```
It seems to break up each entry by the presence of a new `Title` and the label and data are separated by ` : ` with spaces.

**Comment by brantje** [comment]
Does EnPass has the ability to add custom fields?

**Comment by rushsteve1** [comment]
Yes. I just went and checked.

**Comment by brantje** [comment]
Installed EnPass, after a few force closes i managed to create a csv export.
Working on a importer now.

**Comment by KeithIMyers** [comment]
If it helps, here is an example of the CSV Export

"Title","Field","Value","Field","Value",.........,"Note"
"Site1","Login","MyUsername","Password","MyPassword,""
"Site2","Username","MyUsername,"Email","MyEMail","Password","SomePasswordGoesHere","URL","https://kmyers.me","TOTP","GoogleTOTPChallenge","Notes"

**Comment by animalillo** [comment]
Could we get a full example file, with all the possible features of EnPass in use? If possible on both types of encoding.

We would only need a few entries, using all the feature of enpass

**Comment by KeithIMyers** [comment]
Sure @animalillo - I will setup an enpass install and propagate it with some dummy data. Give me a few minutes to get this over to you.

**Comment by KeithIMyers** [comment]
[Passman Sample Data from Enpass.zip](https://github.com/nextcloud/passman/files/706059/Passman.Sample.Data.from.Enpass.zip)

I created a new enpass bank and populated it with sample data. I then exported the data and zipped it up. I included the sample bank as well in the zip file (Enpass_Backup_2017-01-14_11-42-32.walletx). The password to import/decrypt this sample bank is "passman". The Enpass_2017-01-14_11-41-18.txt is the text export directly from Enpass.

I created multiple entries with different scenarios to allow you to see the inner workings. This includes a few custom fields and TOTP in a few cases. I made sure at least one entry in each category is "stock" so you can see the basics. 

For some reason I could not see the option to do a CSV export, it may not be present in the portable version of enpass. I will try the installed version a bit later. 

Let me know if you need anything further

**Comment by animalillo** [comment]
@KeithIMyers Thank you for the files!
Currently we are going to be focusing on getting v2 stable, and this importer will be added for v2.1 hopefully.
I've took a quick look at the files and, as a lot of reverse engineering would be needed to know the encoding, encryption methods and other details, as well as possible changes on future updated of their password manager of the walletx format we would use the CSV and/or the TXT exports from EnPass.

The export you've posted seems good to me, if you can manage to get a CSV export like this one too, it would be perfect!

Thank you!

**Comment by KeithIMyers** [comment]
Just wanted to give a small update for @animalillo - It seems that enPass is currently pushing an update to allow for secure attachments (a feature already in passman). I will be updating to the beta this week and will pull a set of new files with a few attachments so you can see if it is something that will work with passman.

**Comment by animalillo** [comment]
Thank you, it will be great to try to import everything available into passman @KeithIMyers

**Comment by brantje** [comment]
@KeithIMyers 
Import is working
![image](https://cloud.githubusercontent.com/assets/2758767/22934627/e4c3a106-f2cf-11e6-9ced-12d2df8cf079.png)

Can you describe the steps to create an export? 
The custom fields in EnPass i copy to custom fields in passman, however there is no way of telling what field must be secret or not. Eg cvc code.

**Comment by brantje** [comment]
Feature is implemented in the `v2.1.0` branch, if you want you can test it.

**Comment by KeithIMyers** [comment]
This is great news, I am very excited to see this happen

> Can you describe the steps to create an export?

Of course:

1. Access your Enpass Database
2. Select "File" > "Export" > "As Text" (or as CSV depending on how PassMan will want to handle it)
3. Enter your "Master Password" and you will be prompted for a location to save the database. This is an unencrypted database so it should be removed as soon as you import it into Passman

> The custom fields in EnPass i copy to custom fields in passman, however there is no way of telling what field must be secret or not. Eg cvc code.
I do not see how this is marked in the export however one simple solution would be to maintain a list of common sensitive fields (cvc, credit card, etc). The user can mark other sensitive fields as needed.

**Comment by brantje** [comment]
Thanks, i've added the info.
At this moment the following fields will be made secret:
`var secret_fields = ['cvc', 'pin', 'security answer']`

**Comment by KeithIMyers** [comment]
@brantje 
I just tested this in the V2.1.0 Branch and it went well. There is one issue however where the TOTP field seems to be creating a custom field (with the TOTP Secret) instead of importing it as the OTP Secret in Passman. I had to manually copy the value into the OTP secret on Passman. Other then that, it worked perfectly

**Comment by brantje** [comment]
Ouch, missed that one. Will define that as a special field, after we're done uploading the Android app source to GH.

**Comment by KeithIMyers** [comment]
> Ouch, missed that one. Will define that as a special field, after we're done uploading the Android app source to GH.

Speaking of which @brantje - I just installed it a few minutes ago and it has a lot of potential. Still basic as you mentioned but I can honestly see this as a real contender to other password managers.

**Comment by animalillo** [comment]
:D glad you like the app! It's been a lot of hard work lately! I have just uploaded all the app code into github.

**Comment by brantje** [comment]
Thanks, you're welcome to join our mattermost (you can use the gitlab single sign on, from there sign in with your github account)
https://mattermost.passman.cc/signup_user_complete/?id=549o395qr7rxzk8pwje8e81cty

You are our first alpha tester :confetti_ball:

**Comment by brantje** [comment]
Fixed in 2.1

**Comment by brantje** [comment]
@KeithIMyers TOTP fields are now detected on import. Added in 46ed638378a2123dff7d51fb91224b4a258e5bf9
